### PR TITLE
first functions to get the zone id for a dns name

### PR DESCRIPTION
**Pull request by JorritSalverda**
Please leave comments on best practices around

- file structure
- package structure
- error handling
- variable naming
- function naming
- anything else that makes sense in the golang world but doesn't elsewhere

So far I haven't brought in any external packages except for the golang ones in order to figure out what goes on below the hood. Particularly for interpreting cli commands and flags this sucks :)

I'd also like to know how to be able to test the code in cloudflare.go by 'stubbing' the http calls and responses. And possibly directly testing the non-public methods for edge cases.

### CODE REVIEW

File: cloudflare.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+// GetCloudflareZoneByDNSName returns the Cloudflare zone by looking it up with a dnsName, possibly including subdomains; also works for TLDs like .co.uk
+func GetCloudflareZoneByDNSName(dnsName, authKey, authEmail string) (CloudflareZone, error) {
+	var zone CloudflareZone
+
+	// split dnsName
+	dnsNameParts := strings.Split(dnsName, ".")
+
+	// verify dnsName has enough parts
+	if len(dnsNameParts) < 3 {
```
Comment (etiennetremel) [inline cloudflare.go]: should this be 2 instead of 3? then it also match domain.com otherwise it only match _www.domain.com_ and _domain.co.uk_

File: cloudflare.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	sourceLength := len(source)
+	copy(destination, source[sourceLength-numberOfItems:])
+
+	return destination, nil
+}
+
+func getCloudflareZoneByName(zoneName, authKey, authEmail string) ([]CloudflareZone, error) {
+
+	// create api url
+	findZoneURI := fmt.Sprintf("https://api.cloudflare.com/client/v4/zones/?name=%v", zoneName)
```
Comment (etiennetremel) [inline cloudflare.go]: Same, I think this is good practice to have this in an Config struct from the main.go or different file

File: cloudflare.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+	defer response.Body.Close()
+
+	json.NewDecoder(response.Body).Decode(&body)
+
+	return body.Zones, nil
+}
+
+// GetCloudflareZoneByDNSName returns the Cloudflare zone by looking it up with a dnsName, possibly including subdomains; also works for TLDs like .co.uk
+func GetCloudflareZoneByDNSName(dnsName, authKey, authEmail string) (CloudflareZone, error) {
```
Comment (etiennetremel) [inline cloudflare.go]: I'm not sure about this way of handling DNS zone, I'm maybe picky but I kind of prefer the steps they use in the [cloud-flare-go](https://github.com/cloudflare/cloudflare-go/blob/master/cloudflare.go#L54): get the full list of zones and make a simple condition to return the matching one

File: cloudflare.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	destination := make([]string, numberOfItems)
+	sourceLength := len(source)
+	copy(destination, source[sourceLength-numberOfItems:])
+
+	return destination, nil
+}
+
+func getCloudflareZoneByName(zoneName, authKey, authEmail string) ([]CloudflareZone, error) {
```
Comment (etiennetremel) [inline cloudflare.go]: I think it's better to have this authKey/authEmail outside of the argument, and use a Config struct that get defined in the main.go

File: cloudflare_test.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+func TestGetLastItemsFromSliceReturnsErrorWhenSourceIsNil(t *testing.T) {
+
+	_, err := getLastItemsFromSlice(nil, 4)
+
+	if err == nil {
+		t.Error("Expected error")
+	}
+}
+
+func TestGetLastItemsFromSliceReturnsErrorWhenNrOfItemsIsLargerThanNrOfSourceItems(t *testing.T) {
```
Comment (etiennetremel) [inline cloudflare_test.go]: that's a long name :smile:

File: cloudflare.go (hunk 2 above)
Comment (JorritSalverda) [inline cloudflare.go]: Wrapping authKey and authEmail in a struct is easy, but the cloudflare api base url not so much, because you can't provide a default value for a property in a struct or use a const in a struct. So how would you do that?

File: cloudflare.go (hunk 3 above)
Comment (JorritSalverda) [inline cloudflare.go]: What I'll do is search by name in combination with match all. If the number of returned results is larger than 1 page I'll refine the search and otherwise I can find the correct zone in the results.

With their approach you run the risk of having to go through multiple pages of zones if you search for let's say co.uk.

File: cloudflare.go (hunk 1 above)
Comment (JorritSalverda) [inline cloudflare.go]: Smaller than seems completely incorrect to check whether there's enough parts :)

Comment (JorritSalverda) [inline cloudflare.go]: DnsName is ZoneName plus a dns record. But in this case it can't be used for setting the @ record indeed.

**Comment by markvincze** [comment]
So this would replace some of the stuff we currently have in shell scripts?

**Comment by markvincze** [comment]
If this gets more extended in the future, it might be interesting to look at a CLI framework, for example [cobra](https://github.com/spf13/cobra) looks pretty cool.

**Comment by JorritSalverda** [comment]
Idea is to run this as an application in any of our kubernetes clusters and watch for services with annotations. If new ones are added, existing ones change or at a regular interval dns records will be updated.

After this one I want to build separate but comparable apps for letsencrypt certificates - so we can renew without redeploying -, vault secrets and internal tls keys. This moves complexity from the deploy scripts into these apps where it can be well tested and run independent from deployments.

**Comment by JorritSalverda** [comment]
I definitely want to use a CLI library like cobra or kingpin, but to get the hang of the golang basics I started out by doing it manually.

File: cloudflare.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+	}
+	if numberOfItems > len(source) {
+		return nil, fmt.Errorf("main: argument numberOfItems (%v) is larger than number of items in argument source (%v)", numberOfItems, len(source))
+	}
+
+	destination := make([]string, numberOfItems)
+	sourceLength := len(source)
+	copy(destination, source[sourceLength-numberOfItems:])
+
+	return destination, nil
```
Comment (GVdP-Travix) [inline cloudflare.go]: Using `copy` creates a proper copy of the slice. If you don't need the exact copy just return the sub-slice. Considering that it's a slice of strings, and strings are immutable, you are likely to have a copy of a slice containing the same pointers.

File: cloudflare.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+	ID     string `json:"id"`
+	Name   string `json:"name"`
+	Status string `json:"status"`
+}
+
+func getLastItemsFromSlice(source []string, numberOfItems int) ([]string, error) {
+
+	// validate arguments
+	if source == nil {
+		return nil, errors.New("main: argument source is nil")
```
Comment (GVdP-Travix) [inline cloudflare.go]: Good practice for errors is to define them ahead and use them accordingly:
`var (
 ErrEmptySource = errors.New("Empty source")
 ErrTooManyItems = errors.New("Too many times in call") // or whatever.. doesn't have to be that precise
)

func xxxx (src []string, num int) ([]string, error) {
....
return nil, ErrEmptySource
....
}`

This way a user of the function can compare the error.
`x, err:=getLastItemsFromSlice(source, num)
if err==ErrEmptySource {
...
}
`

File: cloudflare.go [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+type CloudflareZone struct {
+	ID     string `json:"id"`
+	Name   string `json:"name"`
+	Status string `json:"status"`
+}
+
+func getLastItemsFromSlice(source []string, numberOfItems int) ([]string, error) {
+
+	// validate arguments
+	if source == nil {
```
Comment (GVdP-Travix) [inline cloudflare.go]: Don't check if a slice is nil, rather check it's length:
`if len(source)==0 {`

A nil slice will always have length zero, but a zero length slice might or might not be nil.

File: cloudflare.go [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+func getCloudflareZoneByName(zoneName, authKey, authEmail string) ([]CloudflareZone, error) {
+
+	// create api url
+	findZoneURI := fmt.Sprintf("https://api.cloudflare.com/client/v4/zones/?name=%v", zoneName)
+
+	// create client in order to set headers
+	client := &http.Client{}
+	request, err := http.NewRequest("GET", findZoneURI, nil)
```
Comment (GVdP-Travix) [inline cloudflare.go]: If you are using Go 1.7, make use of `context` and put the request in a timeout:
`    import (
     ...
     "context"
     "time"
     ...
    )
    
    func getCloudFlareZoneByName(......)... {
     ctx, cancel := context.WithCancel(context.Background())  
     timer := time.AfterFunc(5*time.Second, func() {  // Or whatever length you want
      cancel()
     })
     ....
     ....
     req.WithContext(ctx)
     ....
    }

File: cloudflare_test.go [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+	for i := range a {
+		if a[i] != b[i] {
+			return false
+		}
+	}
+
+	return true
+}
+
+func TestGetLastItemsFromSliceReturns2ItemsWhenSourceIsLongerAndNrOfItemsIs2(t *testing.T) {
```
Comment (GVdP-Travix) [inline cloudflare_test.go]: For testing in different ways the same function, you should make use of sub tests rather than different functions
```
func TestFoo(t *testing.T) {
    // <setup code>
    t.Run("A=1", func(t *testing.T) { ... })
    t.Run("A=2", func(t *testing.T) { ... })
    t.Run("B=1", func(t *testing.T) { ... })
    // <tear-down code>
}```

File: main.go [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+	"flag"
+	"fmt"
+	"log"
+	"os"
+)
+
+func main() {
+
+	// get flag values
+	authKeyPtr := flag.String("authkey", "", "--authkey=r2kjepva04hijzv18u3e9ntphs79kctdxxj5w")
```
Comment (GVdP-Travix) [inline main.go]: For `flag`s more complex than simple settings, I suggest using [urfave's cli](https://github.com/urfave/cli)
It's simple and has no dependencies (or at least, it didn't last time I used it).

**Comment by GVdP-Travix** [comment]
I'm not 100% sold on the cobra cli package, or even the one I use most often, urfave/cli. I find them overbearing and bent on imposing their way of doing things, and end up with way too much code for simple operations. In the end the code becomes less readable.
Don't have a solution, just mentioning.

**Comment by JorritSalverda** [comment]
This particular application won't be a real cli, as it will continuously run, but I've added the possibility to run it with different commands for testing purposes. Will see what works best in the future. The current way of looking at flags and parameters is far from ideal as the order in which the parameters are specified is very important, so not user-friendly at all.

**Comment by JorritSalverda** [comment]
Another option besides cobra and urfave is https://github.com/alecthomas/kingpin.

File: cloudflare.go (hunk 3 above)
Comment (JorritSalverda) [inline cloudflare.go]: Hm, you can't really do a wildcard search it seems, for something like budgetair.co.uk it still does 2 requests. The alternative to get all pages of zones will cost many more requests though.

Btw the cloudflare go library doesn't do this correctly, it still has the following todo

`// TODO: Paginate here. We only grab the first page of results.`
